# Clase 3: ciclos
Notebook diseñado por los docentes Andres Castro y Leonardo Duque.
Modificado por Cristian Guarnizo.

### 1. Estructuras repetitivas (ciclos):

- <font color="#130eb3">__While__</font>
- <font color="#130eb3">__For__</font>


# Estructuras repetitivas


## Ciclo While

Se utiliza para repetir un conjunto de operaciones o instrucciones mientras una condición sea verdadera. La sintaxis en Python es la siguente:

<br>

<font color="#130eb3" size=5>__while__</font> <font size=5>(condición)</font> <font color="blue" size=5>__:__</font>

$\qquad$ instrucción 1
    
$\qquad$ instrucción 2
    
$\qquad$ ...
    
$\qquad$ instrucción n




In [ ]:
a = 10
#Iteración
while (a >= 0): #mientras a sea mayor que cero
    print(a, end= " ") #instruccion 1
    a = a - 1 #instruccion 2

10 9 8 7 6 5 4 3 2 1 0 

In [ ]:
a = 0
while(a <= 10):
    print(a, end=' ')
    a = a + 1

0 1 2 3 4 5 6 7 8 9 10 


## 1.2. Ciclo For

Se utiliza para iterar sobre un conjunto de elementos que hagan parte de un objeto iterable. La sintaxis en Python es la siguente:

<br>

<font color="#130eb3" size=5>__for__ </font>
<font size=5> elemento </font>
<font color="#130eb3" size=5>__in__ </font>
<font size=5> objeto_iterable </font>
<font color="#130eb3" size=5>__:__</font>

$\qquad$ instrucción 1

$\qquad$ instrucción 2

$\qquad$ ...

$\qquad$ instrucción n

En la sintaxis anterior, __objeto_iterable__ se refiere a algun objeto de Python que contenga una serie de elementos, por ejemplo, una cadena de caracteres o un arreglo de datos

In [ ]:
frase = "Lorem ipsum dolor sit amet."

i = 0

for letra in frase: #por cada caracter del texto
    print(f"Letra en la posición {i} es {letra}")
    #print("Letra en la posición ", i, " es ", letra)
    i = i+1

print(f"Frase tiene {len(frase)} letras")

Letra en la posición 0 es L
Letra en la posición 1 es o
Letra en la posición 2 es r
Letra en la posición 3 es e
Letra en la posición 4 es m
Letra en la posición 5 es  
Letra en la posición 6 es i
Letra en la posición 7 es p
Letra en la posición 8 es s
Letra en la posición 9 es u
Letra en la posición 10 es m
Letra en la posición 11 es  
Letra en la posición 12 es d
Letra en la posición 13 es o
Letra en la posición 14 es l
Letra en la posición 15 es o
Letra en la posición 16 es r
Letra en la posición 17 es  
Letra en la posición 18 es s
Letra en la posición 19 es i
Letra en la posición 20 es t
Letra en la posición 21 es  
Letra en la posición 22 es a
Letra en la posición 23 es m
Letra en la posición 24 es e
Letra en la posición 25 es t
Letra en la posición 26 es .
Frase tiene 27 letras


In [ ]:
i=0
while (i<=26):
  print(f"Letra en la posición {i} es {frase[i]}")
  i = i+1

Letra en la posición 0 es L
Letra en la posición 1 es o
Letra en la posición 2 es r
Letra en la posición 3 es e
Letra en la posición 4 es m
Letra en la posición 5 es  
Letra en la posición 6 es i
Letra en la posición 7 es p
Letra en la posición 8 es s
Letra en la posición 9 es u
Letra en la posición 10 es m
Letra en la posición 11 es  
Letra en la posición 12 es d
Letra en la posición 13 es o
Letra en la posición 14 es l
Letra en la posición 15 es o
Letra en la posición 16 es r
Letra en la posición 17 es  
Letra en la posición 18 es s
Letra en la posición 19 es i
Letra en la posición 20 es t
Letra en la posición 21 es  
Letra en la posición 22 es a
Letra en la posición 23 es m
Letra en la posición 24 es e
Letra en la posición 25 es t
Letra en la posición 26 es .


### for + range

Combinando __for__ y __range__ tenemos un for que itera en los valores dados:

En el caso de que deseemos realizar un bucle o ciclo definido en un rango $[a, b)$, podemos utilizar la función <font color="#130eb3">range</font>, la cual nos retorna un objeto iterable en dicho rango:

In [ ]:
help(range)

Help on class range in module builtins:

class range(object)
 |  range(stop) -> range object
 |  range(start, stop[, step]) -> range object
 |  
 |  Return an object that produces a sequence of integers from start (inclusive)
 |  to stop (exclusive) by step.  range(i, j) produces i, i+1, i+2, ..., j-1.
 |  start defaults to 0, and stop is omitted!  range(4) produces 0, 1, 2, 3.
 |  These are exactly the valid indices for a list of 4 elements.
 |  When step is given, it specifies the increment (or decrement).
 |  
 |  Methods defined here:
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash

In [ ]:
for i in range(0,5,2): #for(int i=0; i<5; i=i+2){}
    print(i)

0
2
4


In [ ]:
list(rango)

[0,
 2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 94,
 96,
 98,
 100,
 102,
 104,
 106,
 108,
 110,
 112,
 114,
 116,
 118,
 120,
 122,
 124,
 126,
 128,
 130,
 132,
 134,
 136,
 138,
 140,
 142,
 144,
 146,
 148,
 150,
 152,
 154,
 156,
 158,
 160,
 162,
 164,
 166,
 168,
 170,
 172,
 174,
 176,
 178,
 180,
 182,
 184,
 186,
 188,
 190,
 192,
 194,
 196,
 198,
 200,
 202,
 204,
 206,
 208,
 210,
 212,
 214,
 216,
 218,
 220,
 222,
 224,
 226,
 228,
 230,
 232,
 234,
 236,
 238,
 240,
 242,
 244,
 246,
 248,
 250,
 252,
 254,
 256,
 258,
 260,
 262,
 264,
 266,
 268,
 270,
 272,
 274,
 276,
 278,
 280,
 282,
 284,
 286,
 288,
 290,
 292,
 294,
 296,
 298,
 300,
 302,
 304,
 306,
 308,
 310,
 312,
 314,
 316,
 318,
 320,
 322,
 324,
 326,
 328,
 330,
 332,
 334,
 336,
 338,
 340,
 342,
 344,
 346,
 348,
 350,

In [ ]:
for var in range(10,35):
    print(var,end=' ')

10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 

También podemos utilizar la sintaxis range$(inicio, final-1, paso)$:

In [ ]:
a = -11 #desde 1
b = 11 #hasta 11 - 1 = 10
paso = 2 #de 2 en 2

for a in range(a,b+1,paso):
    print(a, end = " ")



-11 -9 -7 -5 -3 -1 1 3 5 7 9 11 

In [ ]:
a = 1 #desde 1
b = 11 #hasta 11 - 1 = 10
paso = 2 #de 2 en 2

rango = range(a, b, paso)
print(rango)
print(type(rango))
for i in rango:
    print(f"Elemento: {i}")

range(1, 11, 2)
<class 'range'>
Elemento: 1
Elemento: 3
Elemento: 5
Elemento: 7
Elemento: 9


In [ ]:
#imprimir únicamente lo múltiplos de 3 desde 1 hasta 50
for i in range(3,51,3):
    print(i, end=' ')


3 6 9 12 15 18 21 24 27 30 33 36 39 42 45 48 

In [ ]:
for i in range(1,51,1):
    if(i%3==0):
        print(i, end=' ')

3 6 9 12 15 18 21 24 27 30 33 36 39 42 45 48 

In [ ]:
#For que cuente desde 20 hasta -10 en pasos de 5
for i in range(20,-11,-1):
    print(i, end=' ')

20 19 18 17 16 15 14 13 12 11 10 9 8 7 6 5 4 3 2 1 0 -1 -2 -3 -4 -5 -6 -7 -8 -9 -10 


## Ejercicio en clase

Generemos una tabla de conversión de grados Celsius a Farenheit en el rango $[-20, 40]$ con __paso=5__, utilizando ciclo while y for:

## $$\left(\frac{9}{5} °C\right) + 32$$

In [ ]:
#Ciclo While
C = -20
paso = 5

while (C <= 40):
  F = (9/5*C) + 32
  print(f"El valor de {C}º en Farenheit es {F}")
  C += paso

El valor de -20º en Farenheit es -4.0
El valor de -15º en Farenheit es 5.0
El valor de -10º en Farenheit es 14.0
El valor de -5º en Farenheit es 23.0
El valor de 0º en Farenheit es 32.0
El valor de 5º en Farenheit es 41.0
El valor de 10º en Farenheit es 50.0
El valor de 15º en Farenheit es 59.0
El valor de 20º en Farenheit es 68.0
El valor de 25º en Farenheit es 77.0
El valor de 30º en Farenheit es 86.0
El valor de 35º en Farenheit es 95.0
El valor de 40º en Farenheit es 104.0


In [ ]:
#Ciclo for
for C in range(-20,41,5):
  F = (9/5*C) + 32
  print(f"El valor de {C}º en Farenheit es {F}")

El valor de -20º en Farenheit es -4.0
El valor de -15º en Farenheit es 5.0
El valor de -10º en Farenheit es 14.0
El valor de -5º en Farenheit es 23.0
El valor de 0º en Farenheit es 32.0
El valor de 5º en Farenheit es 41.0
El valor de 10º en Farenheit es 50.0
El valor de 15º en Farenheit es 59.0
El valor de 20º en Farenheit es 68.0
El valor de 25º en Farenheit es 77.0
El valor de 30º en Farenheit es 86.0
El valor de 35º en Farenheit es 95.0
El valor de 40º en Farenheit es 104.0


## Ejercicios

### 1.  Crear un programa que permita adivinar un número. Se debe generar un número aleatorio del 1 al 100. A continuación va pidiendo números y va respondiendo si el número a adivinar es mayor o menor que el introducido, además de los intentos que quedan (máximo 10 intentos). El programa termina cuando se acierta el número o cuando se termine el número máximo de intentos.

In [ ]:
## Celda para hacer el ejercicio
import random
numAleatorio = random.randint(0,100)
Intentos = 10
flag = True
while flag:
  print(f"Tienes {Intentos} intentos")
  numIngresado = int(input("Ingrese un numero entero del 0 al 100: "))
  if (numAleatorio > numIngresado):
    print("Numero aleatorio buscado es mayor.")
  elif (numAleatorio < numIngresado):
    print("Numero aleatorio buscado es menor.")
  else:
    print("Adivinaste!!!")
    flag = False
  Intentos -= 1
  if Intentos == 0:
    flag = False
    print("Se acabaron los intentos, mejor suerte la proxima vez.")


Tienes 10 intentos
Ingrese un numero entero del 0 al 100: 1000
Numero aleatorio buscado es menor.
Tienes 9 intentos


KeyboardInterrupt: Interrupted by user

In [ ]:
numAleatorio = random.randint(0,100)
cantidadIntentos = 10
for Intentos in range(cantidadIntentos):
  print(f"Tienes {cantidadIntentos-Intentos} intentos")
  numIngresado = int(input("Ingrese un numero entero del 0 al 100: "))
  if (numAleatorio > numIngresado):
    print("Numero aleatorio buscado es mayor.")
  elif (numAleatorio < numIngresado):
    print("Numero aleatorio buscado es menor.")
  else:
    print("Adivinaste!!!")
    break
else:
    print("Se acabaron los intentos, mejor suerte la proxima vez.")


Tienes 10 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 9 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 8 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 7 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 6 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 5 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 4 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 3 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 2 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Tienes 1 intentos
Ingrese un numero entero del 0 al 100: 1
Numero aleatorio buscado es mayor.
Se acabaron los intentos, mejor suerte la proxima vez.


### 2. Escribir un programa que solicite ingresar números hasta que se introduzca un cero. Debe imprimir la suma y el promedio de todos los números ingresados.

In [ ]:
#ciclo while
con, suma, flag = 0,0,True

while flag:
  numero = int(input("Ingrese un numero: "))
  if numero == 0:
    flag = False
  else:
    con += 1
    suma += numero

if con:
  promedio = suma/con
else:
  promedio = 0

print(f"La suma de los numeros es {suma} y el promedio de los {con} numeros es\
 {promedio} ")

Ingrese un numero: 50
Ingrese un numero: 42
Ingrese un numero: 17
Ingrese un numero: 9
Ingrese un numero: 0
La suma de los numeros es 118 y el promedio de los 4 numeros es 29.5 


50, 42, 17, 9 --> suma = 118, promedio = 29.5

### 3. Escribir un programa que imprima todos los números pares entre dos números que se le pidan al usuario, [A, B] con $$A < B$$

In [ ]:
A = int(input("ingrese un número: "))
B = int(input("ingrese otro número: "))

if A<B:
  for numero in range(A,B+1):
    if numero%2==0:
      print(numero, end= ' ')

ingrese un número: 10
ingrese otro número: 20
10 12 14 16 18 20 

## Control en ciclos
Python cuenta con comandos que permiten alterar el funcionamiento de los ciclos.

| Comando    | Descripción |
| -------- | ------- |
| pass  | No hace nada. Comando utilizado para crear los metodos de las clases.   |
| break  | Rompe el ciclo. Se termina el proceso iterativo.   |
| continue | Salta a la siguiente iteracion, no se ejecuta el codigo que este debajo del Continue dentro del bloque ciclico.  |


In [ ]:
for k in range(10):
  if k == 1:
    continue
  if k <= 1:
    print("Iteracion menor e igual 1.")
  if k == 3:
    pass
  if k == 5:
    break
  print("Iteracion: ",k)

Iteracion menor e igual 1.
Iteracion:  0
Iteracion:  2
Iteracion:  3
Iteracion:  4


Adicionalmente, cuando un ciclo se rompe por medio de `break` no podemos ejecutar el condicional `else`. Pero si el ciclo termina correctamente entra al bloque `else`.

In [ ]:
for k in range(10):
  if k == 1:
    continue
  if k <= 1:
    print("Iteracion menor e igual 1.")
  if k == 3:
    pass
  if k == 5:
    pass
    break
  print("Iteracion: ",k)
else:
  print("Entra al else")

Iteracion menor e igual 1.
Iteracion:  0
Iteracion:  2
Iteracion:  3
Iteracion:  4


In [ ]:
#ciclo while
con, suma, flag = 0,0,True
numero = int(input("Ingrese un numero: "))
numeros = []
while numero:
  con += 1
  suma += numero
  numeros.append(numero)
  numero = int(input("Ingrese un numero: "))

if con:
  promedio = suma/len(numeros)
else:
  promedio = 0
print(f"Los numero ingresado son {numeros}")
print(f"La suma de los numeros es {suma} y el promedio de los {con} numeros es\
 {promedio} ")

Ingrese un numero: 1
Ingrese un numero: 2
Ingrese un numero: 3
Ingrese un numero: 4
Ingrese un numero: 5
Ingrese un numero: 0
Los numero ingresado son [1, 2, 3, 4, 5]
La suma de los numeros es 15 y el promedio de los 5 numeros es 3.0 


In [ ]:
len(numeros)

5

In [ ]:
lista = ['texto',1,1.0,range(4),[[1,2,3],2,3]]

In [ ]:
lista[4][0][1]

2

In [ ]:
tupla = (1,2,3)

In [ ]:
tupla[0] = 2

TypeError: 'tuple' object does not support item assignment

In [ ]:
lista

['texto', 1, 1.0, range(0, 4), [[1, 2, 3], 2, 3]]

In [ ]:
lista[0] = 'Hola'

In [ ]:
lista

['Hola', 1, 1.0, range(0, 4), [[1, 2, 3], 2, 3]]